Tests.Unit.Core.ApplicationServices/ApplicationServices/UIModuleCustomisationTest/UIModuleCustomisationTests.cs

```diff
@@ -0,0 +1,279 @@
+        private readonly Mock<IOrganizationService> _organizationServiceMock;
+        private readonly Mock<IEntityIdentityResolver> _identityResolverMock;
+        private readonly Mock<IUIModuleCustomizationRepository> _organizationRepositoryMock;
+
+        private readonly UIModuleCustomizationService _uiModuleCustomizationService;
+
+        public UIModuleCustomisationTests()
+        {
+            _transactionManagerMock = new Mock<ITransactionManager>();
+            _authorizationContextMock = new Mock<IAuthorizationContext>();
+            _userContextMock = new Mock<IOrganizationalUserContext>();
+            _organizationServiceMock = new Mock<IOrganizationService>();
+            _identityResolverMock = new Mock<IEntityIdentityResolver>();
+            _organizationRepositoryMock = new Mock<IUIModuleCustomizationRepository>();
+
+            _uiModuleCustomizationService = new UIModuleCustomizationService(
+                _transactionManagerMock.Object,
+                _userContextMock.Object,
+                _organizationServiceMock.Object,
+                _identityResolverMock.Object,
+                _organizationRepositoryMock.Object);
+        }
+
+        [Theory]
+        [InlineData("1234")]
+        [InlineData("Test-abc")]
+        [InlineData("Test.123")]
+        [InlineData("123.Test")]
+        [InlineData("Test.?")]
+        [InlineData("?.?")]
+        [InlineData("_")]
+        public void ModuleCustomisation_Returns_BadInput_When_Key_Is_Incorrect(string key)
+        {
+            //arrange
+            var module = A<string>();
+            var organization = new Organization{Id = 1};
+            var nodes = PrepareTestNodes(1, key);
+
+            //act
+            var result = organization.ModifyModuleCustomization(module, nodes);
+
+            //assert
+            Assert.True(result.Failed);
+            Assert.Equal(OperationFailure.BadInput, result.Error.FailureType);
+        }
+        
+        [Fact]
+        public void ModuleCustomisation_Returns_BadInput_When_Nodes_Contain_Duplicates()
+        {
+            //arrange
+            var module = A<string>();
+            var organization = new Organization(){Id = 1};
+            var testKey = "Test.Key";
+            var nodes = PrepareTestNodes(2, testKey);
+
+            //act
+            var result = organization.ModifyModuleCustomization(module, nodes);
+
+            //assert
+            Assert.True(result.Failed);
+            Assert.Equal(OperationFailure.BadInput, result.Error.FailureType);
+        }
+
+        [Fact]
+        public void ModuleCustomisation_Mirrors_The_Nodes()
+        {
+            var module = A<string>();
+            var organization = new Organization{Id = 1};
+            var nodes = PrepareTestNodes();
+            
+            var result = organization.ModifyModuleCustomization(module, nodes);
+
+            Assert.True(result.Ok);
+            Assert.Equal(nodes.Count, result.Value.Nodes.Count);
```

Comment:
testing the count is not enough. you must verify that the state is exactly as you expect including changes